src/docker/shared.rs

```diff
@@ -334,65 +432,138 @@ impl Directories {
         let mount_root = host_root.as_posix_absolute()?;
         let mount_cwd = mount_finder.find_path(cwd, false)?;
 
-        toolchain.set_sysroot(|p| mount_finder.find_mount_path(p));
-
-        // canonicalize these once to avoid syscalls
-        let sysroot_mount_path = toolchain.get_sysroot().as_posix_absolute()?;
-
         Ok((
-            Directories {
-                cargo,
-                xargo,
+            PackageDirectories {
                 target: metadata.target_directory.clone(),
-                nix_store,
                 host_root,
                 mount_root,
                 mount_cwd,
-                toolchain,
-                cargo_mount_path,
-                xargo_mount_path,
-                sysroot_mount_path,
             },
             metadata,
         ))
     }
+}
+
+#[derive(Debug)]
+pub struct Directories {
+    toolchain: ToolchainDirectories,
+    package: PackageDirectories,
+}
```

Comment:
maybe also the methods should have a doc comment about what it returns

Reply:
Is it a good idea to deref the general `Directories` in `ToolchainDirectories`? One encapsulates 2 very different things.

Reply:
The issue I see is that there's two places to add a function in, `*Directories` and then `Directories` to defer.

The easiest way to solve part of this would be a `Deref` impl to deref to the most common one. I think it's a good idea to avoid some pain, but there is an issue with that though, which is that that's not really what `Deref` is meant to be used for (however in practice it is)

Reply:
Another possibility is inside we could just immediately create a reference to `toolchain_dirs` and use that.

Reply:
this is what I'm imagining with the deref impl

```rust
impl std::ops::Deref for Directories {
    type Target = ToolchainDirectories;
    fn deref(&self) -> &Self::Target {
        &self.toolchain
    }
}

impl Directories {
    pub fn assemble(
        mount_finder: &MountFinder,
        metadata: CargoMetadata,
        cwd: &Path,
        toolchain: QualifiedToolchain,
    ) -> Result<(Self, CargoMetadata)> {
        let (package, metadata) = PackageDirectories::assemble(mount_finder, metadata, cwd)?;
        let toolchain = ToolchainDirectories::assemble(mount_finder, toolchain)?;

        Ok((Directories { toolchain, package }, metadata))
    }

    pub fn toolchain_directories(&self) -> &ToolchainDirectories {
        &self.toolchain
    }

    pub fn package_directories(&self) -> &PackageDirectories {
        &self.package
    }

    pub fn target(&self) -> &Path {
        &self.package.target
    }

    pub fn host_root(&self) -> &Path {
        &self.package.host_root
    }

    pub fn mount_root(&self) -> &str {
        &self.package.mount_root
    }

    pub fn mount_cwd(&self) -> &str {
        &self.package.mount_cwd
    }
}
```

Reply:
I've changed it so we use `package_dirs` and `toolchain_dirs` separately, which means we can remove the redundant code above without having to have slightly-changed semantics after derefing.